### PR TITLE
feat(auth): Google sign-in for invited members

### PR DESCRIPTION
**Pull request by tinnhlt-gra**
## What

Lets an already-invited `app_user` sign in with Google, auto-linked by verified email, reusing the existing Gmail-capture Google OAuth client (no new operator credentials).

## Contract

Two new browser-navigation operations, both public (pre-auth), added to `backend/api/crm.yaml`:
- `GET /auth/oidc/{provider}/start` — redirects to Google's consent screen (authorization code + PKCE S256).
- `GET /auth/oidc/{provider}/callback` — completes the round trip, sets `crm_session` on success, redirects to a neutral failure notice otherwise.

`provider` is enumerated to `google` only today.

## Mechanics

- `internal/compose/oidcverify.go`'s existing Google JWKS verifier (built for the Gmail Pub/Sub push) is generalized to return claims and take an injected identity check, so the login flow and the push webhook share one hand-rolled RS256 verifier with no new dependency.
- A new `federated_identity` table maps `(provider, subject) -> user_id`. Resolution tries the linked subject first (re-checking the account is still live before trusting it), falls back to email for a first link, and **never creates an account** — an unrecognized or non-live email is refused exactly like an unrecognized password login.
- The signed state/PKCE cookie (`internal/compose/oidcloginstate.go`) mirrors the connector OAuth flow's own state-signing mechanics but is domain-separated by a type tag, so a token minted for one flow can never be mistaken for the other even when both share the same operator key.
- `identity` never imports `compose`; every cross-module edge (verifier, exchanger, state signer) is injected from `compose/googlesignin.go`, mirroring `WithGmailCapture`'s self-gating shape exactly.
- Google's `redirect_uri` and the post-round-trip SPA landing deliberately read different bases (`--api-base-url` vs `--public-base-url`), matching the existing connector flow's own split for a deployment where the api and the SPA are different origins.
- Frontend: `startFederatedSignIn` is now a real `location.assign` to the start route; a neutral "didn't work" notice is read once from the callback's failure marker and scrubbed from the address, one-shot, surviving React StrictMode's double-invoked lazy initializer correctly (pure read, side-effecting clear moved to an effect).

## Tests

- `internal/compose`: JWKS verifier claims/rejection table, PKCE/state signer round-trip + tamper/expiry/cross-flow rejection, token exchange (happy path, missing token, non-OK status, unreachable endpoint, malformed JSON), Google sign-in wiring (absent/complete config, split-origin redirect bases).
- `internal/modules/identity`: unit tests for the handlers' pure-Go edge cases (unconfigured provider, missing cookie, state mismatch, exchange/verify failure); integration tests over a real Postgres for subject-first resolution, email fallback, refusing an unknown email, refusing a **deactivated but linked** account, subject re-link on email recycling, and the full HTTP round trip (success and refusal).
- `cmd/api`: boot-time redirect-base validation and the split-origin redirect_uri/landing selection.
- Frontend: the redirect, the failure notice (shown once, scrubbed, never replayed on a later unrelated remount), and the design-review preview switch (never live-navigates).
- New/changed files all clear >80% coverage.
- `make check`, `make test-integration` (0 skips), and the whole frontend `vitest run` are green; `craft static --strict` reports 0 blocker/major/minor.

## UAT

No real Google OAuth credentials available in this environment (same limitation the earlier closed attempt at this feature, #337, declared). Verified against a real dev stack: absence-gating (no button when unconfigured), the unconfigured routes 404, and the neutral failure-notice path end to end (shown once, address scrubbed, not replayed on reload) — GIF attached below. UAT caught and fixed a real bug (the StrictMode double-invoke issue above) before it reached review.

Manual/UAT notes: `.tmp/google-oidc-login/MANUAL-TEST.md` (local, not tracked).

## Review

Reviewed by an independent Fable pass and a Codex pass over the full branch diff; both surfaced real findings, all adjudicated and either fixed (re-check liveness on a linked subject, split-origin redirect bases, the StrictMode notice-replay bug, the preview build's live-navigation bug, the login/connector state cross-parse gap, a duplicated entropy helper) or parked with a recorded rationale (OIDC nonce as defense-in-depth only given PKCE+backchannel exchange; two lower-priority efficiency-only findings).

## Not verified here

The real Google round-trip — the actual redirect to `accounts.google.com`, a human completing consent, and the token exchange + ID-token verification against Google's real JWKS. No Google OAuth app credentials are available in this environment. The operator supplies the OAuth client; the redirect URI to register follows the repo's existing api-origin convention (`<api-base-url>/v1/auth/oidc/google/callback`).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---
## Summary by cubic
Invited members could only sign in with a password; now they can sign in with Google, auto-linked by verified email, reusing the Gmail-capture OAuth client so no new operator credentials are needed. The new `/auth/oidc/google/start` and callback routes handle the OAuth round trip, refuse unknown, deactivated, or never-activated emails, and never create an account; when unconfigured, the button is hidden and the routes answer 404.

**New Features**
- Links identities in a `federated_identity` table and falls back to verified-email linking on first sign-in, re-checking the account is still live before trusting a linked subject.
- Refuses consent denials, blank email/subject claims, and any non-live or unactivated email; refusals are rate-limited per IP, logged to `system_log`.
- The one-shot state cookie clears only after state verification succeeds, so a forged cross-site link can't cancel someone else's pending sign-in.
- OIDC routes bypass auth for GET only; log truncation backs off to a UTF-8 boundary, and the UI-preview "unavailable" marker can't disable a provider an installation actually serves.

**Migration**
- Register `<api-base-url>/v1/auth/oidc/google/callback` as a Google Cloud Console redirect URI on the Gmail-capture OAuth client.
- Boot requires `--api-base-url` to be a bare origin when sign-in is enabled, so a malformed value fails fast instead of reaching Google.

<sup>Written for commit 8c7ea71abd69e1693520ddae6f87a78d9e60835e. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/margince/margince/pull/3217?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

## Summary by CodeRabbit

* **New Features**
  * Added Google federated sign-in using secure OIDC authorization-code PKCE.
  * Authentication capabilities now report configured sign-in providers.
  * Existing active accounts can sign in or link their Google identity.
  * Added localized, neutral messaging when federated sign-in fails.

* **Bug Fixes**
  * Improved Google identity and email verification.
  * Added protection against excessive federated sign-in attempts.
  * Refused unknown, inactive, unverified, or unsupported accounts.
  * Prevented unavailable sign-in options from triggering navigation or network requests.

* **Documentation**
  * Documented Google sign-in redirect URI configuration requirements.

### CODE REVIEW

**Comment by tinnhlt-gra** [comment]
## UAT proof

No real Google OAuth credentials in this environment, so this covers what can be verified without one: absence-gating (no button when unconfigured), the unconfigured `/auth/oidc/google/start` 404ing, and the neutral failure-notice path end to end — shown once on the callback's failure redirect, the address scrubbed in the same paint, and not replayed on a later reload. This exact test caught and fixed a real bug (a React StrictMode double-invoke dropping the notice) before it reached review — see the PR description.

google-signin-uat

File: backend/cmd/api/googlesignin.go [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+// mcp.connector_enabled, and an installation that runs Google sign-in
+// without the MCP connector would otherwise bake a malformed or
+// credential-bearing base straight into the redirect_uri this sends to
+// Google — a boot-time refusal is the only place that can catch it before
+// it reaches a request Google's own server logs. --api-base-url is not
+// separately validated for the same reason gmail.go/graph.go's own APIBaseURL
+// fields are not: it is an operator-supplied override of an already-validated
+// origin, not a second untrusted input.
+func googleSignInOptions(cfg apiConfig, stdout io.Writer) ([]compose.Option, error) {
+	redirectBase := cfg.apiBaseURL
```
[the remark this answers is not in the document]
Comment (tinnhlt-gra) [reply backend/cmd/api/googlesignin.go:44]: Fixed via docs, not code: added a note to docs/reference/configuration.md under --api-base-url explaining that the Gmail-capture client's authorized redirect URIs need this callback added in the Google Cloud Console. Not something the code itself can validate or enforce.

File: frontend/src/i18n/en.ts [hunk 2]
```diff
@@ -4815,6 +4815,8 @@ export const en = {
   "auth.noticeSignedOut": "You have been signed out.",
   "auth.noticeSessionExpired":
     "Your session expired. Sign in again to continue.",
+  "auth.noticeOidcFailed":
+    "Sign-in with Google didn't work — try again or use your password.",
```
[the remark this answers is not in the document]
Comment (tinnhlt-gra) [reply frontend/src/i18n/en.ts]: Fixed: removed the em dash from auth.noticeOidcFailed in en.ts, de.ts, and vi.ts (split into two sentences).

File: backend/internal/modules/identity/ssologin.go [hunk 3]
```diff
@@ -0,0 +1,461 @@
...
+// (error 22021) — a mid-rune cut would fail that write silently (this is a
+// best-effort audit trail) and lose the very refusal record the split
+// exists to keep, precisely for the attacker-controlled multi-byte value
+// this function exists to bound.
+func truncateForLog(s string, n int) string {
+	if len(s) <= n {
+		return s
+	}
+	cut := n
+	for cut > 0 && !utf8.RuneStart(s[cut]) {
```
[the remark this answers is not in the document]
Comment (tinnhlt-gra) [reply backend/internal/modules/identity/ssologin.go:102]: Fixed in 8c7ea71ab: clearLoginStateCookie now runs only after stateSigner.Verify confirms provider+nonce genuinely match, not on mere cookie presence. Added TestOidcSignInCallbackStateMismatchNeverClearsTheCookie as the regression test.

**Comment by tinnhlt-gra** [comment]
## Post-review hardening

Two review rounds (Fable/Codex before opening, then CodeRabbit + cubic across five follow-up pushes) surfaced real defects beyond the original implementation, each fixed with a dedicated regression test:

- **HIGH**: an invited-never-activated member (or the agent seat) had no `password_hash` check in the email-fallback lookup, making that email a permanent, token-free login the moment anyone controlled it on Google.
- A forged cross-site link could clear a victim's real, unrelated, still-pending sign-in's state cookie (SameSite=Lax rides a top-level GET) — fixed by only clearing the cookie once provider+nonce genuinely verify.
- `truncateForLog` could split a UTF-8 rune, silently failing the `system_log` write it exists to make possible.
- The preview-build switch could make a real installation's real OIDC provider buttons permanently inert if ever combined with one.
- Plus: `ON DELETE CASCADE` on `federated_identity.user_id`, a redundant index removed, `--api-base-url` validated, the OIDC `error` query param (consent denial) now handled, and several test/doc accuracy fixes.

All backend/frontend suites, `make check`, and `craft static --strict` are green on the final commit; CodeRabbit and cubic both report no further findings. The UAT proof above still reflects the shipped behavior — none of the fixes changed the observable login/failure UX.
